### PR TITLE
added productization stuff & debugged ... 

### PR DESCRIPTION
**Pull request by przygienda**
* added unit test
* removed couple bugs
* added thread local string on filter
* clarified documentation

### CODE REVIEW

**Comment by przygienda** [comment]
Don't pull up yet. I realized that if the filter & the record keys are in different orders the whole thing doesn't work.

**Comment by przygienda** [comment]
OK, that should be all kosher now. Please look over. I tried to build the flyweight hashmap upfront and just .clone() that on every serializer but failed ...

**Comment by przygienda** [comment]
somehow 1.15 barks on the test ...

File: src/lib.rs [hunk 1]
```diff
@@ -54,68 +65,198 @@ type KVFilterList = Vec<(String, HashSet<String>)>;
 /// The filtering map can contain further key "packet" and value "send".
 /// With that the output for "receive" would be filtered.
 ///
-/// More precisely, a key is ignored until present and an entry must
-/// match for all the keys present any of the values to be left
-/// unfiltered. Refer to the test for an example.
+/// More precisely
```
Comment (dpc) [inline src/lib.rs:68]: IIRC, markdown won't render the below as a list, until there's an empty line between.

**Comment by dpc** [comment]
Generally ACK.

So, I am a tab-guy myself, but in Rust I am sticking to the standard formatting generated by `rustfmt` everywhere, so please just run `cargo fmt` on the repo just to keep it consistent.

I was originally planning to go with ordered matching. `slog` is supposed to preserve the order of KV-pairs so it should be possible to know the exact order beforehand. This would increase the performance quite a bit, I believe (no benchmarking done though).

However, since this whole crate is pretty much created to address your issue, it is entirely up to you to decide here. If you think figuring out the exact order is not practical/not worth it, it's fine with me.

I'll review this more carefully sometime soon (today/tomorrow).

**Comment by przygienda** [comment]
Agree on possible perf impact. But it's to fragile to expect the user to "order the values" IMO. If you wouldn't have forced the "serialize" interface in 2.0 we could have gotten the (k,v) in a single shot and do far more efficient comparison than one-by-one-key. I was also thinking about the hashing but we can't since on info!(log, ...) the logger may change and with it the K,V so we'd need to cache location+logger hash for the stuff to work correctly. 

I will reformat and resubmit.

**Comment by dpc** [comment]
I agree about the fragility.

I don't understand the "forced the "serialize" interface in 2.0". Can you explain?

Yeah, the caching won't be easy in the general case.

Also, I might have mentioned it before, but I think it might be really useful to you, so I want to make sure you've considered it. In `slog 2.0` `Logger` itself implements `Drain`. It's a `Drain` that will forward whatever was logged into it as it was logged directly on it with `info!(...)`.

In slog v1, you had one `Drain` and all `Logger`s using it (the whole hierarchy created from it) when used in `info!(...)` and similar statements, would call its `log` to handle the logging decisions. So you had to do all the filtering etc. in one place, losing the knowledge where did it come from. 

```
Drain (all filters etc) <- Logger <- sub-Logger <- sub-Logger
```
 
In slog v2 you can split a big hierarchy, and put some logging decision somewhere in between.

```
Drain <- Logger <- sub-Logger Drain (with som local filters) <- sub-Logger
```

You do that, by building a new `Drain` wraping an existing `Logger` and creating new hierarchy with `Logger::root()`.

This way, if you pass a given `Logger` to a whole subsystem, thread or a given part of your code, you can do some filtering only on log records coming from that part. Eg. `kvfilter` would have to go through part of the keys, or you could control filtering based on loging level for each part separately.

It is a bit more involved (since at the compile time, you need to divide your logging hierarchy into pieces and put filtering in between), but if you're going for performance, and pieces of your software are already clear (eg. each spawned thread, each subsystem), it might be well worth it.

**Review by dpc — APPROVED** [review]

**Comment by przygienda** [comment]
1. serialize, I mean I can't access the KV hierarchy now except serialize interface, i.e. emit_...
2. yes, will remember that. Don't think it's applicable since certain keywords are sprinkled everywhere, e.g. "FSM" and the same key may have multiple values both on filter as well as log statement 
3. thanks for folding in

**Comment by dpc** [comment]
Let me know if I should just release  as 0.3

**Comment by przygienda** [comment]
yes, please, I'll add another pull up. Only brushed minor detail on doc.

**Pull request merged**